packages/core/src/components/tabs2/tabs.tsx

```diff
@@ -0,0 +1,285 @@
+import { generateTabPanelId, generateTabTitleId, TabTitle } from "./tabTitle";
+
+// <Tabs id="tabs">
+//     <Tab id="a" title="Alpha">
+//         <h1>first panel</h1>
+//     </Tab>
+//     <Tab id="b" title="Beta">
+//         <h1>second panel</h1>
+//     </Tab>
+//     <input type="text" placeholder="Search..." />
+// </Tabs>
+
+// TODO
+// vertical key bindings? up/dn
+
+type TabElement = React.ReactElement<ITab2Props & { children: React.ReactNode }>;
+
+const TAB_SELECTOR = `.${Classes.TAB}`;
+
+export interface ITabs2Props extends IProps {
+    /**
+     * Whether the selected tab indicator should animate its movement.
+     * @default true
+     */
+    animate?: boolean;
+
+    /**
+     * Initial selected tab `id`. Note that this prop refers only to `<Tab>` children;
+     * other types of elements are ignored.
+     * @default first tab
+     */
+    defaultSelectedTabId?: TabId;
+
+    /**
+     * Unique identifier for this `Tabs` container. This will be combined with the `id` of each
+     * `Tab` child to generate ARIA accessibility attributes. IDs are required and should be
+     * unique on the page to support server-side rendering.
+     */
+    id: TabId;
+
+    /**
+     * Whether inactive tab panels should be removed from the DOM and unmounted in React.
+     * This can be a helpful performance enhancement but requires careful support for
+     * unmounting and remounting properly.
+     * @default false
+     */
+    renderActiveTabPanelOnly?: boolean;
+
+    /**
+     * Controls selected tab `id`. Providing this prop will put the component in controlled mode.
+     * Unknown ids will result in empty selection (no errors).
+     */
+    selectedTabId?: TabId;
+
+    /**
+     * Whether to show tabs stacked vertically on the left side.
+     * @default false
+     */
+    vertical?: boolean;
+
+    /**
+     * A callback function that is invoked when tabs in the tab list are clicked.
+     */
+    onChange?(newTabId: TabId, prevTabId: TabId): void;
+}
+
+export interface ITabs2State {
+    indicatorWrapperStyle?: React.CSSProperties;
+    selectedTabId?: TabId;
+}
+
+@PureRender
+export class Tabs2 extends AbstractComponent<ITabs2Props, ITabs2State> {
+    public static Tab = Tab2;
+
+    public static defaultProps: ITabs2Props = {
+        animate: true,
+        id: "",
+        renderActiveTabPanelOnly: false,
+        vertical: false,
+    };
+
+    public displayName = "Blueprint.Tabs2";
+
+    private tabElement: HTMLDivElement;
+
+    constructor(props?: ITabs2Props, context?: any) {
+        super(props, context);
+        let selectedTabId = props.selectedTabId;
+        if (selectedTabId === undefined) {
+            // select first tab in absence of user input
+            selectedTabId = props.defaultSelectedTabId === undefined
+                ? this.getTabChildren()[0].props.id
+                : props.defaultSelectedTabId;
+            // NOTE: providing an unknown ID will hide the selection
+        }
```

Comment:
nit: For this kind of controlled/default value juggling, I've started to use a private function called `getInitialX`—in this case, `getInitialSelectedId`—to avoid having a mutable variable and to reduce complexity in the constructor.

```
private getInitialSelectedId = () => {
  const controlledSelectedTabId = props.selectedTabId;
  const { defaultSelectedTabId } = props;

  if (controlledSelectedTabId !== undefined) {
    return controlledSelectedTabId;
  } else if (defaultSelectedTabId !== undefined) {
    return defaultSelectedTabId;
  } else {
    return this.getTabChildren()[0].props.id;
  }
}
```

ubernit: I'd also prefer a `this.getTabChild(index: number)` function instead of `this.getTabChildren()[index: number]`, even though it would only be needed here.